python/ray/air/_internal/torch_utils.py

```diff
@@ -14,42 +14,50 @@ def get_devices() -> List[torch.device]:
     """Gets the correct torch device list configured for this process.
 
     Returns a list of torch CUDA devices allocated for the current worker.
-    If no GPUs are assigned, then it returns a list with a single CPU device.
+    If no accelerators are assigned, then it returns a list with a single CPU device.
 
-    Assumes that `CUDA_VISIBLE_DEVICES` is set and is a
-    superset of the `ray.get_gpu_ids()`.
+    Assumes that `CUDA_VISIBLE_DEVICES`, `ONEAPI_DEVICE_SELECTOR`,
+    `ROCR_VISIBLE_DEVICES`, `NEURON_RT_VISIBLE_CORES`, `TPU_VISIBLE_CHIPS`,
+    `HABANA_VISIBLE_MODULES`,... is set and is a superset of the
+    `ray.get_runtime_context().get_accelerator_ids()`.
     """
     if torch.cuda.is_available():
-        # GPU IDs are assigned by Ray after you specify "use_gpu"
-        # GPU `ray.get_gpu_ids()` may return ints or may return strings.
-        # We should always convert to strings.
-        gpu_ids = [str(id) for id in ray.get_gpu_ids()]
-
         device_ids = []
 
-        if len(gpu_ids) > 0:
-            cuda_visible_str = os.environ.get("CUDA_VISIBLE_DEVICES", "")
-            if cuda_visible_str and cuda_visible_str != "NoDevFiles":
-                cuda_visible_list = cuda_visible_str.split(",")
-            else:
-                cuda_visible_list = []
-
-            # By default, there should only be one GPU ID if `use_gpu=True`.
-            # If there are multiple GPUs, return a list of devices.
-            # If using fractional GPUs, these IDs are not guaranteed
-            # to be unique across different processes.
-            for gpu_id in gpu_ids:
-                try:
-                    device_ids.append(cuda_visible_list.index(gpu_id))
-                except IndexError:
-                    raise RuntimeError(
-                        "CUDA_VISIBLE_DEVICES set incorrectly. "
-                        f"Got {cuda_visible_str}, expected to include {gpu_id}. "
-                        "Did you override the `CUDA_VISIBLE_DEVICES` environment"
-                        " variable? If not, please help file an issue on Github."
-                    )
+        for resource_name, accelerator_ids in (
```

Comment:
If I understand you correctly, since all keys (resource_name) returned by get_runtime_context().get_accelerator_ids() are unique, it will then only ever append available devices for the first accelerator type found. It this the intended behavior?

Reply:
The edge case is, when a single actor has resources of multiple accelerator types, e.g. (GPU [0], HPU [0]), we will have 2 0s in the result device_id list.